### PR TITLE
Only copy output resource to PVC for storage/git output

### PR DESCRIPTION
**Pull request by afrittoli**
At the moment the output_resource module unconditionally adds a
copy step for every output resource in the taskrun.
However not every output resource will generate content to be
copied. Limit the copy to Storage and Git outoputs for now.

This is meant as a hotfix for bug #401 until an implementation
of the image output resource is available via #216. Until
image resource output is implemented, any consumer won't
have guarantee that the image has been actually built and pushed,
and won't know the hash of the created image.

### CODE REVIEW

**Comment by afrittoli** [comment]
> Thank you so much for adding this PR. Can you add test in output resources package that image outputs does not create output related steps?

Sure I can do that. I didn't add tests so far because - I think - this is going away once image output resource is fully implemented, but it's better to have the test and remove it :)

**Comment by shashwathi** [comment]
>  this is going away once image output resource is fully implemented, but it's better to have the test and remove it 

Yes I agree with that too

File: pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go [hunk 1]
```diff
@@ -105,27 +105,34 @@ func AddOutputResources(
...
-					},
-					VolumeMounts: []corev1.VolumeMount{getPvcMount(pipelineRunpvcName)},
-				}}...)
-			}
-			b.Spec.Steps = append(b.Spec.Steps, newSteps...)
+		// Workaround for issue #401. Unless all resource types are implemented as
+		// outputs, or until we have metadata on the resource that declares whether
+		// the output should be copied to the PVC, only copy git and storage output
+		// resources.
+		if (resource.Spec.Type == v1alpha1.PipelineResourceTypeStorage ||
```
Comment (shashwathi) [inline pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go]: Suggestion: 
Nested `if` condition is long and hard to follow. I would recommend adding a function like `allowedOutputResources(string) bool` which will validate type `Storage` and `Git` and return boolean. So you can update `if` condition to be like 

```
// copy resources if its supported list and PVC is present
if allowedOutputResource(resource.Spec.Type) && pipelineRunpvcName != "" {
...
}
```

This makes the code more readable.

Comment (afrittoli) [reply pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go]: ok

Comment (afrittoli) [reply pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go]: Done in an extra commit

File: pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go [hunk 2]
```diff
@@ -186,3 +190,13 @@ func addStoreUploadStep(build *buildv1alpha1.Build,
 	build.Spec.Steps = append(build.Spec.Steps, buildSteps...)
 	return nil
 }
+
+// allowedOutputResource checks if an output resource type produces
+// an output that should be copied to the PVC
+func allowedOutputResource(resourceType v1alpha1.PipelineResourceType) bool {
```
Comment (shashwathi) [inline pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go]: `allowedOutputResources` is not going to change so makes sense to define it in variable section. Something like 
```
var (
// allowedOutputResource checks if an output resource type produces
// an output that should be copied to the PVC
	allowedOutputResources = map[v1alpha1.PipelineResourceType]bool{
		v1alpha1.PipelineResourceTypeStorage: true,
		v1alpha1.PipelineResourceTypeGit:     true,
	}
)
```

and update the if condition to be 
```
if allowedOutputResources[resource.Spec.Type] && pipelineRunpvcName != "" {
}
```

Code is simple and does not need a function. WDYT?

File: pkg/reconciler/v1alpha1/taskrun/resources/output_resource_test.go [hunk 3]
```diff
@@ -716,3 +763,30 @@ func Test_InValid_OutputResources(t *testing.T) {
...
+		desc: "git type is allowed",
+		resourceType: v1alpha1.PipelineResourceTypeGit,
+		expectedAllowed: true,
+	}, {
+		desc: "anything else is not allowed",
+		resourceType: "fooType",
+		expectedAllowed: false,
+	}} {
+		t.Run(c.desc, func(t *testing.T) {
+			actual := allowedOutputResource(c.resourceType)
```
Comment (shashwathi) [inline pkg/reconciler/v1alpha1/taskrun/resources/output_resource_test.go]: Test assertion can be combined into single like 

```
if c.expectedAllowed != allowedOutputResources[c.resourceType] {
  t.Fatalf("Test AllowedOutputResource %s expected %t but got %t", c.desc, c.expectedAllowed, allowedOutputResources[c.resourceType])}
```

File: pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go (hunk 2 above)
Comment (afrittoli) [reply pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go]: I made a function because you asked for it to begin with :)

I'm not sure that list is not going to change... other resources will have output to be copied, like logs of the build process for the image resource.
I hope eventually that list will go away for good and it will be replaced by a proper interface on the resource to declare whether they have output to be copied or not. 

The advantage of having a function is also that it's possible to write a simple unit test for it, but it's true that there's little value in that unit test, so I'm ok with dropping the function if you prefer so.

File: pkg/reconciler/v1alpha1/taskrun/resources/output_resource_test.go (hunk 3 above)
Comment (shashwathi) [reply pkg/reconciler/v1alpha1/taskrun/resources/output_resource_test.go]: If test error contains the expected result and current result then debugging test failures is much easier.

Comment (afrittoli) [reply pkg/reconciler/v1alpha1/taskrun/resources/output_resource_test.go]: Good point, thanks

File: pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go (hunk 2 above)
Comment (shashwathi) [reply pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go]: > I made a function because you asked for it to begin with :)

Sorry for the confusion. My intention was to make the `if` condition simpler and easier to read so I jumped at function as solution. I should have left up to you come up with simpler solution instead of influencing my solution on you. After I saw your solution it felt declaring map in variables section seemed a lot simpler than function

Comment (shashwathi) [reply pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go]: > The advantage of having a function is also that it's possible to write a simple unit test for it, but it's true that there's little value in that unit test, so I'm ok with dropping the function if you prefer so.

You can still have the tests. I think it holds value in showing supported output types explicitly. I like the tests. I have suggested below how to update tests to use map instead of function.

Comment (afrittoli) [reply pkg/reconciler/v1alpha1/taskrun/resources/output_resource.go]: OK, addressed everything - hopefully - in the version :)

**Comment by bobcatfish** [comment]
Seems like a reasonable approach to me! I'll leave it to @shashwathi for the final sign off tho :D

**Comment by shashwathi** [comment]
@afrittoli  Thank you so much for your patience on addressing all my comments. 👍 I appreciate that.
